---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


In [1]:
# this is your first code cell, add more cells for code and mardown as needed
import pandas as pd
import numpy as np

df1 = pd.read_csv("data/sessions.csv")
df2 = pd.read_csv("data/streamers.csv")
df3 = pd.read_csv("data/streams.csv")
df4 = pd.read_csv("data/viewers.csv")

datasets = [df1, df2, df3, df4]

## Exploration

## DF1: Sessions

In [2]:
df1.dtypes

session_id              str
viewer_id               str
streamer_id             str
stream_id               str
started_at              str
ended_at                str
duration_mins         int64
chat_messages_sent    int64
bits_cheered          int64
followed_during        bool
subscribed_during      bool
dtype: object

In [3]:
print(df1.shape)
print(df1.describe())

(22825, 11)
       duration_mins  chat_messages_sent  bits_cheered
count   22825.000000        22825.000000  22825.000000
mean       59.051303            6.442629    398.425323
std        60.383979           46.576309   1571.528658
min         0.000000            0.000000      0.000000
25%        15.000000            0.000000      0.000000
50%        39.000000            1.000000      0.000000
75%        83.000000            5.000000      0.000000
max       443.000000         1987.000000  10000.000000


In [4]:
print(df1['session_id'].nunique())
print(df1['streamer_id'].nunique())
print(df1['stream_id'].nunique())
print(df1['viewer_id'].nunique())

22825
76
1234
1000


I observe there is a relationship between the datasets with Df1 being the base.

In [5]:
df1.head(3)

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False


## DF2: Streamers

In [6]:
df2.dtypes

streamer_id                 str
streamer_name               str
category                    str
language                    str
partner_status             bool
total_followers           int64
avg_concurrent_viewers    int64
dtype: object

In [7]:
print(df2.shape)
df2.head(3)

(76, 7)


,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
0,STR0001,xQc,Just Chatting,English,True,11500000,138706
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
2,STR0003,HasanAbi,Just Chatting,English,True,2600000,34220


## DF3: Streams

In [8]:
df3.dtypes

stream_id                  str
streamer_id                str
started_at                 str
ended_at                   str
stream_duration_hrs    float64
category                   str
peak_viewers           float64
title_has_hype_word       bool
was_raid                  bool
dtype: object

In [9]:
print(df3.shape)
df3.head(3)

(1199, 9)


,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False


## DF4: Viewers

In [10]:
df4.dtypes

viewer_id               str
age_group               str
country                 str
account_age_days      int64
subscription_tier       str
preferred_category      str
dtype: object

In [11]:
print(df4.shape)
df4.head(3)

(1000, 6)


,viewer_id,age_group,country,account_age_days,subscription_tier,preferred_category
0,VWR00001,18-24,AR,355,free,IRL
1,VWR00002,25-34,CA,1670,free,Creative
2,VWR00003,35-44,US,1108,free,Just Chatting


### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


## Cleaning Pipeline

### ID Column Cleaning
Ensures all ID columns are strings, stripped of spaces, and consistently formatted

In [12]:
id_columns = ["session_id", "viewer_id", "streamer_id", "stream_id"]

def clean_ids(df):
    df = df.copy()
    for col in id_columns:
        if col in df.columns:
            df[col] = df[col].astype(str).str.strip()
    return df

In [13]:
datasets = [clean_ids(df) for df in datasets]

df1, df2, df3, df4 = datasets

### Convert Date Columns

In [14]:
def clean_dates(df):
    df = df.copy()
    for col in ["started_at", "ended_at"]:
        if col in df.columns:
            df[col] = pd.to_datetime(df[col], errors="coerce")
    return df

df1 = clean_dates(df1)
df3 = clean_dates(df3)

### => Validate Date Conversion Cleaning
Try to catch weird rows where recorded duration doesn't match timestamps

In [15]:
# Sessions Calculations
df1["calculated_duration_mins"] = (
    df1["ended_at"] - df1["started_at"]
).dt.total_seconds() / 60

df1.head(3)

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during,calculated_duration_mins
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False,72.0
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False,150.0
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False,102.0


In [16]:
# Streams Calculations
df3["calculated_stream_duration_hrs"] = (
    df3["ended_at"] - df3["started_at"]
).dt.total_seconds() / 3600

df3.head(3)

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid,calculated_stream_duration_hrs
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True,3.0
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False,3.0
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False,7.0


## Check Duplicate Entries

In [17]:
def check_key_duplicates(df, key, name):
    duplicates = df.duplicated(subset=key).sum()
    print(f"{name}: {duplicates} duplicate rows by {key}")

check_key_duplicates(df1, ["session_id"], "Dataset 1")
check_key_duplicates(df2, ["streamer_id"], "Dataset 2")
check_key_duplicates(df3, ["stream_id", "streamer_id"], "Dataset 3") # Check if one steam id can appear under mult streamers
check_key_duplicates(df4, ["viewer_id"], "Dataset 4")

Dataset 1: 0 duplicate rows by ['session_id']
Dataset 2: 0 duplicate rows by ['streamer_id']
Dataset 3: 0 duplicate rows by ['stream_id', 'streamer_id']
Dataset 4: 0 duplicate rows by ['viewer_id']


## Clean Categorical Columns

In [18]:
def clean_categories(df):
    df = df.copy()

    categorical_cols = [
        "category",
        "language",
        "age_group",
        "country",
        "subscription_tier",
        "preferred_category"
    ]

    for col in categorical_cols:
        if col in df.columns:
            df[col] = df[col].astype(str).str.strip()

    return df

df3 = clean_categories(df3)
df2 = clean_categories(df2)
df4 = clean_categories(df4)

## Validate Numeric Columns

In [19]:
numeric_checks = {
    "duration_mins": df1,
    "chat_messages_sent": df1,
    "bits_cheered": df1,
    "total_followers": df2,
    "avg_concurrent_viewers": df2,
    "stream_duration_hrs": df3,
    "peak_viewers": df3,
    "account_age_days": df4,
}

def validate(checks):
    for col, df in checks.items():
        print(col)
        print("Missing:", df[col].isna().sum())
        print("Negative:", (df[col] < 0).sum())
        print()

In [20]:
validate(numeric_checks)

duration_mins
Missing: 0
Negative: 0

chat_messages_sent
Missing: 0
Negative: 0

bits_cheered
Missing: 0
Negative: 0

total_followers
Missing: 0
Negative: 5

avg_concurrent_viewers
Missing: 0
Negative: 0

stream_duration_hrs
Missing: 0
Negative: 0

peak_viewers
Missing: 80
Negative: 40

account_age_days
Missing: 0
Negative: 0



## Cleaning Numeric Columns (Drop Entries)

#### Helper functions for cleaning

In [21]:
def drop_invalid_numeric_rows(df, numeric_cols):
    df = df.copy()

    valid_mask = True

    for col in numeric_cols:
        if col in df.columns:
            valid_mask = valid_mask & df[col].notna() & (df[col] >= 0)

    return df[valid_mask]

# Sets invalid values to missing instead of total drop
def replace_invalid_numeric_with_na(df, numeric_cols):
    df = df.copy()

    for col in numeric_cols:
        if col in df.columns:
            df.loc[df[col] < 0, col] = pd.NA

    return df

In [38]:
# Fill missing count-like activity values with 0
df1["chat_messages_sent"] = df1["chat_messages_sent"].fillna(0)
df1["bits_cheered"] = df1["bits_cheered"].fillna(0)

df2["total_followers"] = df2["total_followers"].fillna(0)
df2["avg"] = df2["total_followers"].fillna(0)



# Drop impossible values
df1 = drop_invalid_numeric_rows(
    df1,
    ["duration_mins", "chat_messages_sent", "bits_cheered"]
)

df2 = replace_invalid_numeric_with_na(
    df2,
    ["total_followers", "avg_concurrent_viewers"]
)

df3 = replace_invalid_numeric_with_na(
    df3,
    ["stream_duration_hrs", "peak_viewers"]
)

df4 = drop_invalid_numeric_rows(
    df4,
    ["account_age_days"]
)

og_dataset = datasets

# Update the dataset with cleaned versions
datasets = [df1, df2, df3, df4]

# Preserve Removed Rows
df2_removed = og_dataset[1].loc[~og_dataset[1].index.isin(datasets[1].index)]
df3_removed = og_dataset[2].loc[~og_dataset[2].index.isin(datasets[2].index)]

### Check if Removed IDs appear in Df1

If an entry has been drop, if affects the number of sessions. To mitigate this, I decided on a non-destructive approach by replacing invalid entries with "Missing" to retain the entry.

#### Streamers (Effect)

In [23]:
removed_streamers_used = df1[df1["streamer_id"].isin(df2_removed["streamer_id"])]

print("Sessions affected by removed streamers:", len(removed_streamers_used))
print(removed_streamers_used["streamer_id"].value_counts())

Sessions affected by removed streamers: 0
Series([], Name: count, dtype: int64)


#### Streams (Effect)

In [24]:
removed_streams_used = df1[df1["stream_id"].isin(df3_removed["stream_id"])]

print("Sessions affected by removed streams:", len(removed_streams_used))
print(removed_streams_used["stream_id"].value_counts().head(20))

Sessions affected by removed streams: 0
Series([], Name: count, dtype: int64)


### Post Clean Validation

In [25]:
numeric_checks = {
    "duration_mins": df1,
    "chat_messages_sent": df1,
    "bits_cheered": df1,
    "total_followers": df2,
    "avg_concurrent_viewers": df2,
    "stream_duration_hrs": df3,
    "peak_viewers": df3,
    "account_age_days": df4,
}

In [26]:
validate(numeric_checks)

duration_mins
Missing: 0
Negative: 0

chat_messages_sent
Missing: 0
Negative: 0

bits_cheered
Missing: 0
Negative: 0

total_followers
Missing: 5
Negative: 0

avg_concurrent_viewers
Missing: 0
Negative: 0

stream_duration_hrs
Missing: 0
Negative: 0

peak_viewers
Missing: 120
Negative: 0

account_age_days
Missing: 0
Negative: 0



Updated invalid entries (negative numeric values) to be missing. Instead of dropping.

In [27]:
for i in range(len(og_dataset)):
    print(f"Update: {og_dataset[i].shape} -> {datasets[i].shape}")

Update: (22825, 11) -> (22825, 12)
Update: (76, 7) -> (76, 8)
Update: (1199, 9) -> (1199, 10)
Update: (1000, 6) -> (1000, 6)


Want to retain shape to have 1-1 match with DF1 base.

## Merge Dataset for Exploration

Category inquiry: 
Both Df2 and Df3 have a "category" column.


My proposed approach/assumption:<br>
Dataset 2’s category means the streamer’s usual/current category.<br>
Dataset 3’s category means the category of that specific stream.<br>

In [28]:
df2 = df2.rename(columns={
    "category": "streamer_category"
})

df3 = df3.rename(columns={
    "category": "stream_category",
    "started_at": "stream_started_at",
    "ended_at": "stream_ended_at"
})

Main Activity/session table; using df1 as base

In [29]:
merged = (
    df1
    .merge(df2, on="streamer_id", how="left", validate="many_to_one")
    .merge(df4, on="viewer_id", how="left", validate="many_to_one")
    .merge(df3, on=["stream_id", "streamer_id"], how="left", validate="many_to_one")
)

## Audit Merged Dataset

In [30]:
print("Missing streamer info:", 
        merged["streamer_name"].isna().sum(), 
        f"({round(merged["streamer_name"].isna().sum()/df1.shape[0] * 100, 2)}% of Entries)"
     )
print("Missing viewer info:", 
        merged["age_group"].isna().sum(),
        f"({round(merged["age_group"].isna().sum()/df1.shape[0] * 100)}% of Entries)"
     )
print("Missing stream info:", 
        merged["stream_category"].isna().sum(),
        f"({round(merged["stream_category"].isna().sum()/df1.shape[0] * 100, 2)}% of Entries)"
     )

Missing streamer info: 60 (0.26% of Entries)
Missing viewer info: 0 (0% of Entries)
Missing stream info: 985 (4.32% of Entries)


#### Checking: If stream IDs missing entirely or is due to merge being too strict

In [31]:
test_stream_merge = df1.merge(
    df3,
    on="stream_id",
    how="left",
    suffixes=("", "_stream")
)

print("Missing stream info using stream_id only:")
print(test_stream_merge["stream_category"].isna().sum())

Missing stream info using stream_id only:
929


Note/Observation: DF3 is most likely missing datasets associated with DF1 (with variation about ~6%)

### Final Audit

In [32]:
# Create separate merged datasets for different analysis purposes (could be handy in the future idk)
merged_all = merged.copy()

merged_with_stream_info = merged.dropna(subset=["stream_category"])

merged_with_streamer_info = merged.dropna(subset=["streamer_name"])

merged_complete = merged.dropna(subset=[
    "streamer_name",
    "stream_category"
])

### Cleaning Rationale
Instead of dropping rows with invalid numeric values, invalid entries were replaced with missing values to preserve key relationships across datasets. This reduced missing streamer matches from 1,600 rows to 60 rows and missing stream matches from 3,149 rows to 985 rows after merging. Viewer information matched completely, with 0 missing viewer records. Remaining missing stream metadata was retained where possible, and rows may be excluded only for analyses that specifically require stream-level fields.

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [33]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here. 

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [34]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [35]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [36]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [37]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  